tools/dpgen/generate.go

```diff
@@ -0,0 +1,282 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package main
+
+import (
+	"fmt"
+	"slices"
+	"strings"
+
+	"github.com/cilium/ebpf"
+	"github.com/cilium/ebpf/btf"
+	"github.com/mitchellh/go-wordwrap"
+
+	"github.com/cilium/cilium/pkg/datapath/config"
+
+	"golang.org/x/text/cases"
+	"golang.org/x/text/language"
+)
+
+// Common acronyms to transform into stylized form for Go field names.
+var stylized = map[string]string{
+	"bpf":     "BPF",
+	"lxc":     "LXC",
+	"xdp":     "XDP",
+	"ipv4":    "IPv4",
+	"ipv6":    "IPv6",
+	"nat":     "NAT",
+	"mac":     "MAC",
+	"mtu":     "MTU",
+	"id":      "ID",
+	"ip":      "IP",
+	"netns":   "NetNS",
+	"ipcache": "IPCache",
+}
+
+// varsToStruct generates a Go struct from the configuration variables in the
+// CollectionSpec.
+func varsToStruct(spec *ebpf.CollectionSpec, name, kind, comment string, embeds []string) (string, error) {
+	type field struct {
+		comment  string
+		goName   string
+		cName    string
+		typ      string
+		defValue string
+	}
+
+	kind = "kind:" + kind
+
+	fields := make([]field, 0, len(spec.Variables))
+
+	for n, v := range spec.Variables {
+		// Only consider variables in a specific config section to avoid interfering
+		// with unrelated objects.
+		if v.MapName() != config.Section {
+			continue
+		}
+
+		// DECLARE_CONFIG prefixes the variable name with a well-known prefix to
+		// avoid collisions with other variables with common names.
+		n = strings.TrimPrefix(n, config.ConstantPrefix)
+
+		if v.Type() == nil {
+			return "", fmt.Errorf("variable %s has no type information, was the ELF built without BTF?", n)
+		}
+
+		// Skip variables that don't have the requested kind.
+		tags := v.Type().Tags
+		if !slices.Contains(tags, kind) {
+			continue
+		}
+
+		// Pop the kind tag from the list of tags, we don't want to render it out
+		// to the config struct.
+		tags = slices.DeleteFunc(tags, func(s string) bool { return s == kind })
+
+		if len(tags) == 0 || tags[0] == "" {
+			return "", fmt.Errorf("variable %s has no doc comment", n)
+		}
+
+		typ, err := btfVarGoType(v.Type())
+		if err != nil {
+			return "", fmt.Errorf("variable %s: getting Go type: %w", n, err)
+		}
+
+		comment, err := tagsToComment(tags)
+		if err != nil {
+			return "", fmt.Errorf("variable %s: converting tags to comments: %w", n, err)
+		}
+
+		defValue, err := varGoValue(v)
+		if err != nil {
+			return "", fmt.Errorf("variable %s: getting default Go value: %w", n, err)
+		}
+
+		fields = append(fields, field{comment, camelCase(n), n, typ, fmt.Sprintf("%#v", defValue)})
```

Comment:
Nit: should we format the value as is rather? I'm seeing it would otherwise pick hex (0x5dc) rather than int (1500).

```suggestion
		fields = append(fields, field{comment, camelCase(n), n, typ, fmt.Sprintf("%v", defValue)})
```

Reply:
I picked this because according to https://pkg.go.dev/fmt#hdr-Printing, `%#v` is the only documented way I know of that explicitly mentions emitting 'a Go-syntax representation of the value'. `%v` doesn't mention that, I don't know why.

A follow-up PR will add support for all C data types (arrays, structs, unions), and those frequently contain IP/hw addresses that are easier to read in hex, so I figured it wouldn't be a bad choice.

Reply:
I've run the following simple example:

```go
type exampleStruct struct {
	foo    bool
	bar    int
	foobar string
	barfoo uint16
}

func main() {
	e := exampleStruct{false, 2, "hello", 1500}

	fmt.Printf("%v", e)   // {false 2 hello 1500}
	fmt.Printf("%#v", e)  // main.exampleStruct{foo:false, bar:2, foobar:"hello", barfoo:0x5dc}
}
```

Second option is way better that the 1st one, despite `barfoo uint16` being translated to [hex](https://pkg.go.dev/fmt#hdr-Printing:~:text=uint%2C%20uint8%20etc.%3A%20%20%20%20%20%20%20%20%25d%2C%20%25%23x%20if%20printed%20with%20%25%23v).

> A follow-up PR will add support for all C data types (arrays, structs, unions), and those frequently contain IP/hw addresses that are easier to read in hex, so I figured it wouldn't be a bad choice.

I definitively agree with this, I'm ok to merge this format.
Thanks for the explanation!